File: articles/optimization-models-and-objective-functions/fortran/optimization_allocation_model.f90
program optimization_allocation_model
  implicit none

  integer :: a, b, c, d, feasible_count, candidate_count
  integer :: best_a, best_b, best_c, best_d
  real(8) :: budget, total_cost, total_benefit, best_benefit
  integer :: equity_floor

  budget = 75.0d0
  equity_floor = 1
  feasible_count = 0
  candidate_count = 0
  best_benefit = -1.0d0
  best_a = 0
  best_b = 0
  best_c = 0
  best_d = 0

  do a = 0, 8
    do b = 0, 8
      do c = 0, 8
        do d = 0, 8
          candidate_count = candidate_count + 1
          total_cost = 7.0d0*a + 8.0d0*b + 5.0d0*c + 6.0d0*d
          total_benefit = 11.0d0*a + 13.0d0*b + 8.0d0*c + 10.0d0*d

          if (total_cost <= budget .and. a >= equity_floor .and. b >= equity_floor .and. c >= equity_floor .and. d >= equity_floor) then
            feasible_count = feasible_count + 1
            if (total_benefit > best_benefit) then
              best_benefit = total_benefit
              best_a = a
              best_b = b
              best_c = c
              best_d = d
            end if
          end if

        end do
      end do
    end do
  end do

  print '(A,I0,A,I0,A,F10.2,A,4(I0,1X))', 'fortran candidates=', candidate_count, ' feasible=', feasible_count, ' best_benefit=', best_benefit, ' best_allocation=', best_a, best_b, best_c, best_d
end program optimization_allocation_model

File: articles/optimization-models-and-objective-functions/cpp/optimization_allocation_model.cpp
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

struct Program {
    std::string name;
    double benefit;
    double cost;
    int lower;
    int upper;
};

struct Scenario {
    std::string name;
    double budget;
    int equity_floor;
};

int main() {
    const std::vector<Program> programs = {
        {"housing", 11.0, 7.0, 0, 8},
        {"health", 13.0, 8.0, 0, 8},
        {"transport", 8.0, 5.0, 0, 8},
        {"resilience", 10.0, 6.0, 0, 8}
    };

    const Scenario scenario{"cpp_baseline", 75.0, 1};

    double best_benefit = -1.0;
    std::vector<int> best_allocation;
    int feasible_count = 0;
    int candidate_count = 0;

    for (int a = 0; a <= 8; ++a) {
        for (int b = 0; b <= 8; ++b) {
            for (int c = 0; c <= 8; ++c) {
                for (int d = 0; d <= 8; ++d) {
                    const std::vector<int> x = {a, b, c, d};
                    double total_cost = 0.0;
                    double total_benefit = 0.0;
                    bool equity_ok = true;

                    for (std::size_t i = 0; i < programs.size(); ++i) {
                        total_cost += x[i] * programs[i].cost;
                        total_benefit += x[i] * programs[i].benefit;
                        if (x[i] < scenario.equity_floor) {
                            equity_ok = false;
                        }
                    }

                    ++candidate_count;
                    const bool feasible = total_cost <= scenario.budget && equity_ok;

                    if (feasible) {
                        ++feasible_count;
                        if (total_benefit > best_benefit) {
                            best_benefit = total_benefit;
                            best_allocation = x;
                        }
                    }
                }
            }
        }
    }

    std::cout << std::fixed << std::setprecision(2)
              << "cpp candidates=" << candidate_count
              << " feasible=" << feasible_count
              << " best_benefit=" << best_benefit
              << " best_allocation=[";

    for (std::size_t i = 0; i < best_allocation.size(); ++i) {
        if (i > 0) std::cout << ",";
        std::cout << best_allocation[i];
    }

    std::cout << "]\n";
    return 0;
}
